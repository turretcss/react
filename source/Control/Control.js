import React from 'react'
import classnames from 'classnames'

import { indicatorMap, sizeAliases } from '../constants'

const noop = () => {}

const typeAliases = {
  switch: 'checkbox',
}

const Control = ({
  type = 'checkbox',
  className,
  size,
  disabled,
  name,
  value = false,
  readOnly,
  inline,
  color,
  children,
  text,
  reversed,
  onChange = noop,
  // ...props // TODO: do we need to spread props?
}) => (
  <label
    className={classnames('control', className, type, {
      ['control-' + (sizeAliases[size] || size)]: size,
      ['control-inline']: inline,
      ['control-reversed']: reversed,
      [color]: indicatorMap[color],
      disabled: disabled || readOnly,
    })}>
    <input
      name={name}
      type={typeAliases[type] || type}
      disabled={disabled || readOnly}
      checked={value}
      onClick={event => onChange(event.target.checked)}
      onChange={noop}
    />
    <span className="control-indicator" />
    {(children || text) && <span className="control-label">{children || text}</span>}
  </label>
)

export const Checkbox = Control
export const Radio = props => <Control {...props} type="radio" />
export const Switch = props => <Control {...props} type="switch" />

export default Control
