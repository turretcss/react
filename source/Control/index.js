export { default } from './Control'
export { default as Control, Checkbox, Radio, Switch } from './Control'
