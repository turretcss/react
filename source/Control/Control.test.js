import React from 'react'
import renderer from 'react-test-renderer'

import { Checkbox, Radio, Switch } from './Control'

describe('Control', () => {
  it('renders checkbox', () => {
    const tree = renderer.create(<Checkbox>Hello</Checkbox>).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('renders radio', () => {
    const tree = renderer.create(<Radio>Hello</Radio>).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('renders switch', () => {
    const tree = renderer.create(<Switch>Hello</Switch>).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('renders checkbox (text prop)', () => {
    const tree = renderer.create(<Checkbox text="Hello" />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('prefers children over text prop', () => {
    const tree = renderer.create(<Checkbox text="Ignored">Hello</Checkbox>).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('renders checked checkbox', () => {
    const tree = renderer.create(<Checkbox value={true} text="Hello" onChange={value => (response = value)} />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('renders without span.control-label', () => {
    const tree = renderer.create(<Checkbox />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('renders custom className', () => {
    const tree = renderer.create(<Checkbox className="custom-do-dah" />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('renders disabled', () => {
    const tree = renderer.create(<Checkbox disabled />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('renders disabled (readOnly prop)', () => {
    const tree = renderer.create(<Checkbox readOnly />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('renders name prop', () => {
    const tree = renderer.create(<Checkbox name="terms" />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('renders size prop', () => {
    const tree = renderer.create(<Checkbox size="xl" />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('renders size prop (alias)', () => {
    const tree = renderer.create(<Checkbox size="xlarge" />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('renders color prop (indicator)', () => {
    const tree = renderer.create(<Checkbox color="success" />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('renders inline prop', () => {
    const tree = renderer.create(<Checkbox inline />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('renders reversed prop', () => {
    const tree = renderer.create(<Checkbox reversed />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('renders complex', () => {
    const tree = renderer
      .create(
        <Checkbox
          value={true}
          onChange={value => {}}
          className="thing"
          color="error"
          size="small"
          name="hello"
          text="Chekbox label"
        />
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('calls onChange when input clicked', () => {
    let newValue = null
    const testInstance = renderer.create(<Checkbox onChange={value => (newValue = value)} />).root
    // should execute
    testInstance.findByType('input').props.onClick({ target: { checked: true } })
    // should be ignored
    testInstance.findByType('input').props.onChange({ target: { checked: false } })
    expect(newValue).toEqual(true)
  })
})
