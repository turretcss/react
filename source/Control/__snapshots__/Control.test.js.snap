// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Control prefers children over text prop 1`] = `
<label
  className="control checkbox"
>
  <input
    checked={false}
    disabled={undefined}
    name={undefined}
    onChange={[Function]}
    onClick={[Function]}
    type="checkbox"
  />
  <span
    className="control-indicator"
  />
  <span
    className="control-label"
  >
    Hello
  </span>
</label>
`;

exports[`Control renders checkbox (text prop) 1`] = `
<label
  className="control checkbox"
>
  <input
    checked={false}
    disabled={undefined}
    name={undefined}
    onChange={[Function]}
    onClick={[Function]}
    type="checkbox"
  />
  <span
    className="control-indicator"
  />
  <span
    className="control-label"
  >
    Hello
  </span>
</label>
`;

exports[`Control renders checkbox 1`] = `
<label
  className="control checkbox"
>
  <input
    checked={false}
    disabled={undefined}
    name={undefined}
    onChange={[Function]}
    onClick={[Function]}
    type="checkbox"
  />
  <span
    className="control-indicator"
  />
  <span
    className="control-label"
  >
    Hello
  </span>
</label>
`;

exports[`Control renders checked checkbox 1`] = `
<label
  className="control checkbox"
>
  <input
    checked={true}
    disabled={undefined}
    name={undefined}
    onChange={[Function]}
    onClick={[Function]}
    type="checkbox"
  />
  <span
    className="control-indicator"
  />
  <span
    className="control-label"
  >
    Hello
  </span>
</label>
`;

exports[`Control renders color prop (indicator) 1`] = `
<label
  className="control checkbox success"
>
  <input
    checked={false}
    disabled={undefined}
    name={undefined}
    onChange={[Function]}
    onClick={[Function]}
    type="checkbox"
  />
  <span
    className="control-indicator"
  />
</label>
`;

exports[`Control renders complex 1`] = `
<label
  className="control thing checkbox control-s error"
>
  <input
    checked={true}
    disabled={undefined}
    name="hello"
    onChange={[Function]}
    onClick={[Function]}
    type="checkbox"
  />
  <span
    className="control-indicator"
  />
  <span
    className="control-label"
  >
    Chekbox label
  </span>
</label>
`;

exports[`Control renders custom className 1`] = `
<label
  className="control custom-do-dah checkbox"
>
  <input
    checked={false}
    disabled={undefined}
    name={undefined}
    onChange={[Function]}
    onClick={[Function]}
    type="checkbox"
  />
  <span
    className="control-indicator"
  />
</label>
`;

exports[`Control renders disabled (readOnly prop) 1`] = `
<label
  className="control checkbox"
>
  <input
    checked={false}
    disabled={true}
    name={undefined}
    onChange={[Function]}
    onClick={[Function]}
    type="checkbox"
  />
  <span
    className="control-indicator"
  />
</label>
`;

exports[`Control renders disabled 1`] = `
<label
  className="control checkbox"
>
  <input
    checked={false}
    disabled={true}
    name={undefined}
    onChange={[Function]}
    onClick={[Function]}
    type="checkbox"
  />
  <span
    className="control-indicator"
  />
</label>
`;

exports[`Control renders inline prop 1`] = `
<label
  className="control checkbox control-inline"
>
  <input
    checked={false}
    disabled={undefined}
    name={undefined}
    onChange={[Function]}
    onClick={[Function]}
    type="checkbox"
  />
  <span
    className="control-indicator"
  />
</label>
`;

exports[`Control renders name prop 1`] = `
<label
  className="control checkbox"
>
  <input
    checked={false}
    disabled={undefined}
    name="terms"
    onChange={[Function]}
    onClick={[Function]}
    type="checkbox"
  />
  <span
    className="control-indicator"
  />
</label>
`;

exports[`Control renders radio 1`] = `
<label
  className="control radio"
>
  <input
    checked={false}
    disabled={undefined}
    name={undefined}
    onChange={[Function]}
    onClick={[Function]}
    type="radio"
  />
  <span
    className="control-indicator"
  />
  <span
    className="control-label"
  >
    Hello
  </span>
</label>
`;

exports[`Control renders reversed prop 1`] = `
<label
  className="control checkbox control-reversed"
>
  <input
    checked={false}
    disabled={undefined}
    name={undefined}
    onChange={[Function]}
    onClick={[Function]}
    type="checkbox"
  />
  <span
    className="control-indicator"
  />
</label>
`;

exports[`Control renders size prop (alias) 1`] = `
<label
  className="control checkbox control-xl"
>
  <input
    checked={false}
    disabled={undefined}
    name={undefined}
    onChange={[Function]}
    onClick={[Function]}
    type="checkbox"
  />
  <span
    className="control-indicator"
  />
</label>
`;

exports[`Control renders size prop 1`] = `
<label
  className="control checkbox control-xl"
>
  <input
    checked={false}
    disabled={undefined}
    name={undefined}
    onChange={[Function]}
    onClick={[Function]}
    type="checkbox"
  />
  <span
    className="control-indicator"
  />
</label>
`;

exports[`Control renders switch 1`] = `
<label
  className="control switch"
>
  <input
    checked={false}
    disabled={undefined}
    name={undefined}
    onChange={[Function]}
    onClick={[Function]}
    type="checkbox"
  />
  <span
    className="control-indicator"
  />
  <span
    className="control-label"
  >
    Hello
  </span>
</label>
`;

exports[`Control renders without span.control-label 1`] = `
<label
  className="control checkbox"
>
  <input
    checked={false}
    disabled={undefined}
    name={undefined}
    onChange={[Function]}
    onClick={[Function]}
    type="checkbox"
  />
  <span
    className="control-indicator"
  />
</label>
`;
